File: white-hole/src/Operations/MovieInterfaceOperations.hs
module Operations.MovieInterfaceOperations where

import Database.PostgreSQL.Simple
import Entities.User
import qualified Entities.Movie as M
import Entities.Serie
import Data.Char (toLower, toUpper, isNumber, isAlphaNum, isDigit)
import qualified Entities.Rating as R
import Operations.UserOperations
import Operations.MovieOperations
import Operations.UtilOperations
import System.IO



createMovie :: Connection -> IO ()
createMovie conn = do
    putStrLn ""
    putStrLn "Qual o título do filme?"
    title <- getLine
    if null title then
        putStrLn "O filme deve ter um título!" >> createMovie conn
    else
        putStrLn "Qual a data de lançamento do filme? (yyyy-mm-dd)"

    releaseDate <- getLine
    putStrLn "Qual a duração do filme? (Em minutos)"
    duration <- getLine
    putStrLn "Qual a sinopse do filme?"
    summaryT <- getLine
    putStrLn "Digite o nome dos diretores (separados por vírgula):"
    directors <- getLine
    putStrLn "Digite o nome dos atores (separados por vírgula):"
    actors <- getLine
    let summary = if summaryT == "" then "No comments!" else summaryT

    putStrLn "Digite a quais categorias esse filme pertence (digite os números associados separados por espaços):"
    putStrLn "1 - Ação          8  - Fantasia"
    putStrLn "2 - Suspense      9  - Documentário"
    putStrLn "3 - Romance       10 - Drama"
    putStrLn "4 - Comédia       11 - Anime"
    putStrLn "5 - Terror        12 - Mistério"
    putStrLn "6 - Aventura      13 - Infantil"
    putStrLn "7 - Investigação  14 - Ficção científica"
    putStrLn ""

    categoriesTemp <- getLine
    putStrLn "Quer realmente adicionar esse filme? (s/N)"
    confirmation <- getLine
    if null confirmation || map toLower confirmation == "n" then do
        putStrLn "Ok!"
    else do
        movie <- registerMovie conn title releaseDate duration summary
        addCastingToMovie conn movie (splitItems actors [])
        addDirectorsToMovie conn movie (splitItems directors [])
        verified <- verifyCategories (words categoriesTemp)
        let categories = if verified then convertCategories (words categoriesTemp) [] else []
        addCategoriesToMovie conn movie categories
        putStrLn ""
        putStrLn "Filme cadastrado com sucesso!"


printMoviesList :: [M.Movie] -> Integer -> IO ()
printMoviesList [] _ = return ()
printMoviesList movies seqNum = do
    putStrLn (show seqNum ++ " - " ++ capitalize (M.title (head movies)))
    printMoviesList (tail movies) (seqNum + 1)


tenBestMovies :: Connection -> IO ()
tenBestMovies conn = undefined


tenBestMoviesByCategory :: Connection -> IO ()
tenBestMoviesByCategory conn = undefined


showMovie :: Connection -> User -> M.Movie -> IO ()
showMovie conn user movie = do
    putStrLn "-------------------------------------------------"
    putStrLn (M.title movie)
    putStrLn "-------------------------------------------------"
    putStrLn ("Duração (em minutos): " ++ M.summary movie)
    putStrLn ("Data de lançamento: " ++ M.releaseDate movie)
    categories <- getCategoriesOfMoviesInOneString conn movie
    putStrLn ("Categorias: " ++ categories)
    putStrLn ""
    putStrLn "1 - Mostrar casting do filme."
    putStrLn "2 - Marcar como assistir depois."
    putStrLn "3 - Avaliar filme."
    putStrLn "4 - Mostrar avaliações do filme."
    putStrLn "5 - Voltar."
    putStrLn ""
    putStrLn "Digite a opção desejada: "
    hFlush stdout
    option <- getLine
    case option of
        "1" -> printCasting conn movie >> showMovie conn user movie
        "2" -> addToWatchLaterList conn user movie >> putStrLn "Filme adicionado com sucesso na lista!" >> showMovie conn user movie
        "3" -> newRating conn user movie >> showMovie conn user movie
        "4" -> printRatings conn movie >> showMovie conn user movie
        "5" -> hFlush stdout
        x -> putStrLn "Digite uma opção válida" >> showMovie conn user movie



printCasting :: Connection -> M.Movie -> IO ()
printCasting conn movie = do
    casting <- getCasting conn movie
    if null casting then putStrLn "Não há um Casting cadastrado para esse filme!" else printCasting' casting


printCasting' :: [(String, String)] -> IO ()
printCasting' [] = return ()
printCasting' (x:xs) = do
    putStrLn ("* " ++ fst x ++ " - " ++ capitalize (snd x))
    printCasting' xs


newRating :: Connection -> User -> M.Movie -> IO ()
newRating conn user movie = do
    putStrLn ""
    putStrLn "Dê uma nota para o filme (De 1 a 5): "
    nota <- getLine
    if not (isANumber nota True) then putStrLn "Digite um valor válido na próxima vez!" >> newRating conn user movie else putStrLn ""
    putStrLn "Faça um comentário sobre o filme (Se não quiser, basta apertar enter): "
    commentary <- getLine
    putStrLn ""
    avaluateMovie conn user movie (read nota :: Integer) commentary
    putStrLn "Avaliação feita com sucesso!"

printRatings :: Connection -> M.Movie -> IO ()
printRatings conn movie = do
    ratings <- getRatings conn movie
    if null ratings then putStrLn "Esse filme ainda não foi avaliado por algum usuário!" else printRatings' ratings


printRatings' :: [R.Rating] -> IO ()
printRatings' [] = return ()
printRatings' (x:xs) = do
    putStrLn (R.userEmail x ++ " - " ++ show (R.rating x) ++ " - " ++ R.commentary x)
    printRatings' xs

File: white-hole-pl/src/Operations/movie_interface_operations.pl
:- module(movie_interface_operations, [
        createMovie/1,
        printMoviesList/2,
        tenBestMovies/2,
        tenBestMoviesByCategory/2,
        showMovie/3,
        printCasting/2,
        printCastingAux/1,
        newRating/3,
        printRatings/2,
        printRatingsAux/1
    ]).
:- use_module('./util.pl').
:- use_module('../Db/dbOperations.pl').
:- use_module('./user_operations.pl').
:- use_module('./movie_operations.pl').

createMovie(Connection):-
    writeln(""),
    get_input("Qual o título do filme?", X),
    util:strip(X, Title),
    (Title = "" -> writeln("O filme deve ter um título!"), 
        clear_screen_with_confirmation;
        get_input("Qual a data de lançamento do filme? (yyyy-mm-dd)", Y),
        util:strip(Y, ReleaseDate),
        (ReleaseDate = "" -> writeln("A data deve ser uma data válida!"),
            clear_screen_with_confirmation;
            get_input("Qual a duração do filme? (Em minutos)", Z),
            util:strip(Z, Movies),
            ( Movies \= "", util:isANumber(NMovies, Movies), NMovies < 10000 ->
                get_input("Qual a sinopse do filme?", Sinopse),
                get_input("Digite o nome dos diretores (separados por vírgula):", Directors),
                get_input("Digite o nome dos atores (separados por vírgula):", Actors),
                Summary,
                ( Sinopse = "" -> Summary = "Sem sinopse"; Summary = Sinopse),
                writeln(""),
                writeln("Digite a quais categorias esse filme pertence (digite os números associados separados por espaços):"),
                writeln(""),
                writeln("1 - Ação          8  - Fantasia"),
                writeln("2 - Suspense      9  - Documentário"),
                writeln("3 - Romance       10 - Drama"),
                writeln("4 - Comédia       11 - Anime"),
                writeln("5 - Terror        12 - Mistério"),
                writeln("6 - Aventura      13 - Infantil"),
                writeln("7 - Investigação  14 - Ficção científica"),
                get_input("", A),
                util:strip(A, Categories),
                util:split_string(Categories, " ", " ", SplitedCategories),
                (Categories \= "", util:verify_cartegory(SplitedCategories) -> 
                    get_input("Quer realmente adicionar esse filme? (s/N)", Confirmation),
                    string_lower(Confirmation, LowerCase),
                    ( Confirmation \= "", LowerCase \= "n" -> 
                        writeln(""),
                        writeln("Ok!"),
                        clear_screen_with_confirmation;
                        registerMovie(Connection, Title, ReleaseDate, Duration, Summary, Movie),
                        util:splitItems(Actors, ActorsAdd),
                        addCastingToMovie(Connection, Movie, ActorsAdd),
                        util:splitItems(Directors, DirectorsAdd),
                        addDirectorsToMovie(Connection, Movie, DirectorsAdd),
                        util:convertCategories(SplitedCategories, [], CategoriesConverted),
                        addCategoriesToMovie(Connection, Movie, CategoriesConverted),
                        writeln(""),
                        writeln("Filme cadastrado com sucesso!"),
                        clear_screen_with_confirmation
                    )    
                    ;
                    writeln("É necessário que o filme tenha ao menos uma categoria!"),
                    clear_screen_with_confirmation
                )
                ;
                writeln("Digite uma duração válida"),
                clear_screen_with_confirmation
            )
        ),
    ).

printMoviesList([], _).
printMoviesList([Movie | T], Num):-
    Movie = row(_, Title, _, _, _, _),
    util:capitalize(Title, CapitalizedTitle),
    swritef(Saida, "%w  - %w", [Num, CapitalizedTitle]),
    writeln(Saida),
    NextNum is Num + 1,
    printMoviesList(T, NextNum).

tenBestMovies(Connection, User):-
    writeln(""),
    writeln("Os dez filmes mais bem avaliados no momento são: "),
    writeln(""),
    getTenBestMovies(Connection, Movies),
    printMoviesList(Movies, 1),
    writeln(""),
    get_input("Digite o número do filme para acessá-lo (aperte enter para voltar):", Option),
    length(Movies, SLength),
    (Option = "" -> clear_screen;
        ( util:isANumber(NOption, Option), NOption < SLength + 1, NOption > 0 ->
            clear_screen,
            nth1(NOption, Movies, Movie),
            showMovie(Connection, User, Movie),
            tenBestMovies(Connection, User);
            writeln(""),
            writeln("Digite uma opção válida na próxima vez."),
            clear_screen_with_confirmation
        )  
    ).

tenBestMoviesByCategory(Connection, User):-
    writeln(""),
    writeln("Selecione a categoria que você quer vizualizar:"),
    writeln(""),
    writeln("1 - Ação          8  - Fantasia"),
    writeln("2 - Suspense      9  - Documentário"),
    writeln("3 - Romance       10 - Drama"),
    writeln("4 - Comédia       11 - Anime"),
    writeln("5 - Terror        12 - Mistério"),
    writeln("6 - Aventura      13 - Infantil"),
    writeln("7 - Investigação  14 - Ficção científica"),
    writeln(""),
    writeln("Digite o número da categoria:"),
    get_input("", Option),
    util:strip(Option, OptionTrimmed),
    ( OptionTrimmed \= "", util:isANumber(NOption, OptionTrimmed), NOption < 15, NOption > 0 ->
        clear_screen,
        writeln("Os dez filmes mais bem avaliados dessa categoria no momento são (se a categoria não contiver filmes o suficiente, pode não aparecer dez filmes): "),
        util:convertCategories([NOption], [], Categories),
        getMoviesByCategory(Connection, Categories, Movies),
        writeln(""),
        printMoviesList(Movies, 1),
        writeln(""),
        get_input("Digite o número do filme para acessá-lo (aperte enter para voltar):", Input),
        util:strip(Input, InputTrimmed),

        ( InputTrimmed = "" -> clear_screen;
            InputTrimmed \= "", util:isANumber(NInput, InputTrimmed), NInput < 15, NInput > 0 ->
            clear_screen,
            nth1(NInput, Movies, Movie),
            showMovie(Connection, User, Movie),
            tenBestMoviesByCategory(Connection, User)
        );
        writeln(""),
        writeln("Digite uma opção válida na próxima vez."),
        clear_screen_with_confirmation
    ).

showMovie(Connection, User, Movie):-
    Movie = row(_, Title, RDate, Duration, Summary, Rating),
    util:capitalize(Title, CTitle),
    writeln("-------------------------------------------------"),
    writeln(CTitle),
    writeln("-------------------------------------------------"),
    writeln(""),
    swritef(Sinopse, "Sinopse: %w", [Summary]),
    writeln(Sinopse),
    writeln(""),
    write("Duração (em minutos): "), writeln(Duration),
    writeln("Data de lançamento: "), writeln(RDate),
    getCategoriesOfMoviesInOneString(Connection, Movie, CategoriesString),
    write("Categorias: "), writeln(CategoriesString),
    writeln(""),
    write("Nota geral: "), writeln(Rating),
    writeln(""),
    writeln("1 - Mostrar casting do filme."),
    writeln("2 - Marcar como assistir depois."),
    writeln("3 - Avaliar filme."),
    writeln("4 - Mostrar avaliações do filme."),
    writeln("5 - Voltar."),
    writeln(""),
    get_input("Digite a opção desejada: ", Option),
    ( Option = "1" -> clear_screen, printCasting(Connection, Movie), showMovie(Connection, User, Movie);
    Option = "2" -> clear_screen, movie_operations:addToWatchLaterList(Connection, User, Movie, _), writeln("Filme adicionado com sucesso na lista!"), clear_screen_with_confirmation, showMovie(Connection, User, Movie);
    Option = "3" -> clear_screen, newRating(Connection, User, Movie), movie_operations:updateMovieRating(Connection, Movie, _), showMovie(Connection, User, Movie);
    Option = "4" -> clear_screen, printRatings(Connection, Movie), showMovie(Connection, User, Movie);
    Option = "5" -> clear_screen;
        writeln("Digite uma opção válida"), clear_screen_with_confirmation, showMovie(Connection, User, Movie)
    ).

printCasting(Connection, Movie):-
    getCasting(Connection, Movie, Casting),
    ( Casting = [] -> writeln(""), writeln("Não há um Casting cadastrado para esse filme!"), clear_screen_with_confirmation;
    clear_screen,
    writeln("------------------------------------------------"),
    writeln("                    Casting"),
    writeln("------------------------------------------------"),
    writeln(""),
    printCastingAux(Casting), clear_screen_with_confirmation
    ).

printCastingAux([]).
printCastingAux([ row(X, Y) | T]):-
    util:capitalize(Y, Name),
    write("* "), write(X), write(" - "), writeln(Name),
    printCastingAux(T).

newRating(Connection, User, Movie):-
    writeln(""),
    get_input("Dê uma nota para o filme (De 1 a 5): ", Rating),
    ( util:isANumber(N, Rating), member(N, [1,2,3,4,5]) ->
        writeln(""),
        get_input("Faça um comentário sobre o filme (Se não quiser, basta apertar enter): ", C),
        ( C = "" -> Comment = "Sem comentario" ;
            Comment = C
        ),
        writeln(""),
        get_input("Confirmar avaliação (s/N)", Confirmation),
        string_lower(Confirmation, LowerCase),
        ( Confirmation \= "", LowerCase \= "n" -> 
            movie_operations:avaluateMovie(Connection, User, Movie, Rating, Comment, Conf),
            ( Conf =:= 1 -> clear_screen,
                writeln("Avaliação feita com sucesso!");
                clear_screen,
                writeln("Um mesmo usuário não pode avaliar o mesmo filme duas vezes!")
            )
            ;
            writeln("Ok!" ),
            clear_screen_with_confirmation    
        )
        ;
        writeln("Digite um valor válido na próxima vez!"), newRating(Connection, User, Movie)
    ).


printRatings(Connection, Movie):-
    getRatingsMovies(Connection, Movie, Ratings),
    ( Ratings = [] -> writeln(""), writeln("Esse filme ainda não foi avaliado por algum usuário!"), clear_screen_with_confirmation;
        writeln(""),
        writeln("------------------------------------------------"),
        writeln("                  Avaliações"),
        writeln("------------------------------------------------"),
        writeln(""),
        printRatingsAux(Ratings), clear_screen_with_confirmation
    ).


printRatingsAux([]).
printRatingsAux([row(_, Email, _, Rating, Comment) | T]):-
    write(Email), write(" - "), write(Rating), write(" - "), writeln(Comment),
    printRatingsAux(T).